### PR TITLE
Fix logical difference derived constraint property

### PR DESCRIPTION
**Pull request by dgkimura**
Issue manifested itself in the following query:
```
EXPLAIN INSERT INTO mytable
SELECT a.col FROM mytable a
INNER JOIN (
    SELECT _a.col
    FROM mytable _a,
         mytable _b
    WHERE  (_a.col = _b.col)
    EXCEPT (SELECT col FROM mytable)
) b
ON (a.col = b.col);
```

Which produced the algebrized query:
```
+--CLogicalNAryJoin
   |--CLogicalGet "mytable" ("mytable"), Columns: ["col" (0), ...]
   |--CLogicalDifference Output: ("col" (8)), Input: [("col" (8)), ("col" (24))]
   |  |--CLogicalNAryJoin
   |  |  |--CLogicalGet "mytable" ("mytable"), Columns: ["col" (8), ...]
   |  |  |--CLogicalGet "mytable" ("mytable"), Columns: ["col" (16), ...]
   |  |  +--CScalarCmp (=)
   |  |     |--CScalarIdent "col" (8)
   |  |     +--CScalarIdent "col" (16)
   |  +--CLogicalGet "mytable" ("mytable"), Columns: ["col" (24), ...]
   +--CScalarCmp (=)
      |--CScalarIdent "col" (0)
      +--CScalarIdent "col" (8)
```

And incorrectly generated a preprocessed query that referenced col (16)
outside the scope of the outer CLogicalNAryJoin. Note that only col (8)
was in the output columns of CLogicalDifference:
```
+--CLogicalSelect
   |--CLogicalSelect
   |  |--CLogicalNAryJoin
   |  |  |--CLogicalGet "mytable" ("mytable"), Columns: ["col" (0), ...]
   |  |  |--CLogicalDifference Output: ("col" (8)), Input: [("col" (8)), ("col" (24))]
   |  |  |  |--CLogicalNAryJoin
   |  |  |  |  |--CLogicalGet "mytable" ("mytable"), Columns: ["col" (8), ...]
   |  |  |  |  |--CLogicalGet "mytable" ("mytable"), Columns: ["col" (16), ...]
   |  |  |  |  +--CScalarCmp (=)
   |  |  |  |     |--CScalarIdent "col" (8)
   |  |  |  |     +--CScalarIdent "col" (16)
   |  |  |  +--CLogicalGet "mytable" ("mytable"), Columns: ["col" (24), ...]
   |  |  +--CScalarCmp (=)
   |  |     |--CScalarIdent "col" (0)
   |  |     +--CScalarIdent "col" (8)
   |  +--CScalarBoolOp (EboolopAnd)
   |     |--CScalarCmp (=)
   |     |  |--CScalarIdent "col" (0)
   |     |  +--CScalarIdent "col" (16)
   |     +--CScalarCmp (=)
   |        |--CScalarIdent "col" (8)
   |        +--CScalarIdent "col" (16)
   +--CScalarBoolOp (EboolopAnd)
      |--CScalarCmp (=)
      |  |--CScalarIdent "col" (0)
      |  +--CScalarIdent "col" (16)
      +--CScalarCmp (=)
         |--CScalarIdent "col" (8)
         +--CScalarIdent "col" (16)
```

The reason that the out of scope reference of col (16) happened is
because `CLogicalDifference::DerivePropertyConstraint()` had considered
all the childrens' columns when deriving property constraints. The above
example would include both col (8) and col (16) in the equivalence
class, despite the fact that only col (8) is output from
CLogicalDifference.

The problem is that we later performed invalid transformations using the
incorrect equivalence class with col (16). Those transformations led to
plans that would use the column when we no longer held a reference to it
and thus error: "No plan has been computed for required properties".

Fix is for CLogicalDifference operator to only consider the output
columns to calculate the equivalence class columns.

Co-authored-by: Bhuvnesh Chaudhary <bhuvnesh2703@gmail.com>

### CODE REVIEW

File: src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalDifference.h [hunk 1]
```diff
@@ -77,10 +77,10 @@ class CLogicalDifference : public CLogicalSetOp
 
 	// derive constraint property
 	CPropConstraint *
-	DerivePropertyConstraint(CMemoryPool *,	 //mp,
+	DerivePropertyConstraint(CMemoryPool *mp,
 							 CExpressionHandle &exprhdl) const override
 	{
-		return PpcDeriveConstraintPassThru(exprhdl, 0 /*ulChild*/);
+		return PpcDeriveConstraintSetop(mp, exprhdl, false /*fIntersect*/);
```
Comment (khannaekta) [inline src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalDifference.h:83]: Shouldn't we also call the same function when calling  `DerivePropertyConstraint` for `CLogicalDifferenceAll`?

File: src/test/regress/expected/qp_dml_joins.out [hunk 2]
```diff
@@ -4035,3 +4035,16 @@ ERROR:  multiple updates to a row by the same query is not allowed  (seg0 10.152
 UPDATE dml_heap_p SET a = dml_heap_p.b % 2 FROM dml_heap_r WHERE dml_heap_p.b::int = dml_heap_r.b::int and dml_heap_p.a = dml_heap_r.a;
 --Update on table with composite distribution key
 UPDATE dml_heap_p SET b = (dml_heap_p.b * 1.1)::int FROM dml_heap_r WHERE dml_heap_p.b = dml_heap_r.a and dml_heap_p.b = dml_heap_r.b;
+-- Insert with join and except
+SET optimizer_trace_fallback=on;
+CREATE TABLE dml_heap_int (a int) DISTRIBUTED BY (a);
```
Comment (khannaekta) [inline src/test/regress/expected/qp_dml_joins.out:4040]: we don't validate the results for this query or have a mdp test to validate the plan. I think we should add those too.

File: src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalDifference.h (hunk 1 above)
Comment (dgkimura) [reply src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalDifference.h:83]: Yes, thanks for catching!  Updated with an MDP

File: src/test/regress/expected/qp_dml_joins.out (hunk 2 above)
Comment (dgkimura) [reply src/test/regress/expected/qp_dml_joins.out:4040]: Yeah, thanks for calling that out.  Added MDP tests and regress test output check

File: src/test/regress/expected/qp_dml_joins_optimizer.out [hunk 3]
```diff
@@ -4038,3 +4038,26 @@ ERROR:  multiple updates to a row by the same query is not allowed  (seg0 127.0.
 UPDATE dml_heap_p SET a = dml_heap_p.b % 2 FROM dml_heap_r WHERE dml_heap_p.b::int = dml_heap_r.b::int and dml_heap_p.a = dml_heap_r.a;
 --Update on table with composite distribution key
 UPDATE dml_heap_p SET b = (dml_heap_p.b * 1.1)::int FROM dml_heap_r WHERE dml_heap_p.b = dml_heap_r.a and dml_heap_p.b = dml_heap_r.b;
+-- Insert with join and except
+SET optimizer_trace_fallback=on;
```
Comment (chrishajas) [inline src/test/regress/expected/qp_dml_joins_optimizer.out:4042]: From this point on in the file we'll log tracebacks. Ideally I'd like that to happen in the entire file, but I know that's a larger effort so I'm ok with starting here. I just want to make sure not resetting it is intentional after this query.

**Review by chrishajas — APPROVED** [review]
Looks great! Nice fixes!

**Comment by dgkimura** [comment]
Merged. Thanks for reviews!

**Pull request merged**